arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedReaderBuilder.java

```diff
@@ -129,11 +130,31 @@ public VectorizedReader<?> primitive(
     if (desc.getMaxRepetitionLevel() > 0) {
       return null;
     }
-    Types.NestedField icebergField = icebergSchema.findField(parquetFieldId);
-    if (icebergField == null) {
+    Types.NestedField logicalType = icebergSchema.findField(parquetFieldId);
+    if (logicalType == null) {
       return null;
     }
+
+    Types.NestedField physicalType = logicalType;
+    PrimitiveType.PrimitiveTypeName typeName = primitive.getPrimitiveTypeName();
+    if (OriginalType.DECIMAL.equals(primitive.getOriginalType())) {
+      org.apache.iceberg.types.Type type;
+      if (PrimitiveType.PrimitiveTypeName.INT64.equals(typeName)) {
+        // Use BigIntVector for long backed decimal
+        type = Types.LongType.get();
+      } else if (PrimitiveType.PrimitiveTypeName.INT32.equals(typeName)) {
+        // Use IntVector for int backed decimal
+        type = Types.IntegerType.get();
+      } else {
+        // Use FixedSizeBinaryVector for binary backed decimal
+        type = Types.FixedType.ofLength(primitive.getTypeLength());
+      }
+      physicalType =
+          Types.NestedField.of(
+              logicalType.fieldId(), logicalType.isOptional(), logicalType.name(), type);
```

Comment:
Do we really need this? Figure out the physical type and logical type here and pass them to the reader. 
I understand we need to know the physical type of the decimal fields when constructing an Arrow `Field`, but can we move this step to `VectorizedArrowReader#allocateFieldVector` directly? As is currently done with schema evolution, build the correct Arrow `Field` based on the actual Parquet column type, see `VectorizedArrowReader#allocateVectorBasedOnTypeName`. 
It seems that the change will be much simpler if it can be done.

Reply:
+1, you have access to the column descriptor in VectorizedArrowReader anyway

Reply:
Updated, thanks for the suggestion.